### PR TITLE
fix sorting bug

### PR DESCRIPTION
**Pull request by anautsch**
As described in #1722 , there is an issue with sorting and DDP.

This is to open up a PR. The goal is not fixing all possibilities - the goal is to restore functionality.
A testing is not yet implemented (needs DDP set-up; that might be a different sort of github workflow altogether).

https://github.com/speechbrain/speechbrain/blob/c229dbcde906a850ffcacd7de2c5b02755f2d74e/speechbrain/core.py#L726

defines a shuffle variable that is simply re-used for the flags that were redefined in #1518 ; defaults to `False`, such that sorting is kept if wanted througout.

### CODE REVIEW

**Comment by anautsch** [comment]
> To what is this variable connected? a yaml parameter?

yeah kinda - but it might be in conflict with some other concept beneath.
maybe the idea is bogus. need to test - let's use this as a playground to get sth working

well, actually, the sorting flag is the flag?

**Comment by anautsch** [comment]
just put it to "False" instead; this simply reverts the other PR that caused the bug w/o needing to rethink recipe & hparam logic.

**Comment by anautsch** [comment]
Added a test and an initial fix outline. The test uses a CPU ddp backend, `gloo`.

I run into these nearby-sortings w/ sorting & ddp:
```
tensor([2.2050, 2.3400, 2.6000, 2.8700, 3.1050, 3.1500, 3.2200, 3.2800, 3.4550,
        3.6500, 3.7000, 4.2650, 4.4701, 4.7600, 4.8350, 4.8900, 5.1600, 6.1450,
        1.7750], dtype=torch.float64)

tensor([6.1450, 5.1600, 4.8900, 4.8350, 4.7600, 4.4701, 4.2650, 3.7000, 3.6500,
        3.4550, 3.2800, 3.2200, 3.1500, 3.1050, 2.8700, 2.6000, 2.3400, 2.2050,
        6.2000], dtype=torch.float64)
```

**Comment by anautsch** [comment]
The test needed to be adjusted to DDP & pytorch's DistributedSampler which has:
```
        drop_last (bool, optional): if ``True``, then the sampler will drop the
            tail of the data to make it evenly divisible across the number of
            replicas. If ``False``, the sampler will add extra indices to make
            the data evenly divisible across the replicas. Default: ``False``.
```

Test works for the proposed fix on DDP & sorting. Bc of the drop_last & replicas, the final elements of the last batch won't be sorted within that batch. Yet, DDP seems to distribute the batches w/ replicas BUT
* for sorted (filtered) datasets, the total number of examples does not change
* for random (shuffling), the number of examples increases, as replicas are included.

The sorting asserts are thus tailored to the dummy data (and a bit lazy with "[:-1]" subset testing across all batches, instead of final batch only).

Ready for review.

File: speechbrain/core.py [hunk 1]
```diff
@@ -1200,11 +1211,18 @@ def _wrap_distributed(self):
             for name, module in self.modules.items():
                 if any(p.requires_grad for p in module.parameters()):
                     module = SyncBatchNorm.convert_sync_batchnorm(module)
-                    module = DDP(
-                        module,
-                        device_ids=[self.device],
-                        find_unused_parameters=self.find_unused_parameters,
-                    )
+                    if self.distributed_backend == "gloo":
```
Comment (TParcollet) [inline speechbrain/core.py:1214]: Why has gloo a different behavior now?

File: speechbrain/utils/distributed.py [hunk 2]
```diff
@@ -126,11 +126,12 @@ def ddp_init_group(run_opts):
                 "--distributed_backend=nccl"
             )
         else:
-            if run_opts["local_rank"] + 1 > torch.cuda.device_count():
-                raise ValueError(
-                    "Killing process " + str() + "\n"
-                    "Not enough GPUs available!"
-                )
+            if not run_opts["distributed_backend"] == "gloo":
```
Comment (TParcollet) [inline speechbrain/utils/distributed.py:129]: Again, why is gloo treated like that? :p

**Review by TParcollet — COMMENTED** [review]
See comments.

File: speechbrain/core.py (hunk 1 above)
Comment (AsuMagic) [reply speechbrain/core.py:1214]: I think @anautsch might have assumed that `gloo` is only desirable/used in a CPU DDP scenario. [This is the general recommendation,](https://pytorch.org/docs/stable/distributed.html#which-backend-to-use) but using `gloo` for GPU-based DDP appears to be a supported option, even if the recommendation is `nccl`.

Comment (anautsch) [reply speechbrain/core.py:1214]: `gloo` is the only way for CPU to do DDP - i.e. what it wants, must be there for testing when GPUs are not there (github workflows).

`gloo` has API differences to the other two, as it uses here one CPU node only, and not several GPU nodes. Otherwise, pytorch throws an error - and that's it then.

File: speechbrain/utils/distributed.py (hunk 2 above)
Comment (anautsch) [reply speechbrain/utils/distributed.py:129]: the way it is implemented for GPU is fully incompatible with GPU, in fact: gloo will crash when run in cpu only.

This pytorch example for gloo is incompatible with SB, as it was before the contributed fix
https://pytorch.org/tutorials/intermediate/dist_tuto.html

Comment (anautsch) [reply speechbrain/utils/distributed.py:129]: for this part, if there is only CPU, ofc there are not enough GPUs ;)

File: speechbrain/core.py (hunk 1 above)
Comment (anautsch) [reply speechbrain/core.py:1214]: btw - I did not try to make `mpi` available, compare
https://pytorch.org/docs/stable/distributed.html

Comment (anautsch) [reply speechbrain/core.py:1214]: `gloo` cannot do GPU !!!

Comment (anautsch) [reply speechbrain/core.py:1214]: a DDP integration testing -> not possible w/o GPUs
no idea how to make that workable within github workflows ;)
could be done offline & before another release is published - but then, what should the test case look like?

**Review by TParcollet — APPROVED** [review]
LGTM
